### ISSUE TITLE
Question about samples structure difference between DDPO and DanceGRPO

### ISSUE THREAD

**Issue by ZongxinZhu1** [XueZeyue/DanceGRPO#99]
Thank you very much for your time and the excellent work on DanceGRPO!

I have a question regarding the design choice in the samples construction within DanceGRPO compared to the original DDPO implementation.
In the original DDPO code, the samples are constructed as follows:
samples.append({
    "prompt_ids": prompt_ids,
    "prompt_embeds": prompt_embeds,
    "timesteps": timesteps,
    "latents": latents[:, :-1],              # latent before t
    "next_latents": latents[:, 1:],          # latent after t
    "log_probs": log_probs,
    "rewards": rewards,
})

However, in DanceGRPO / fastvideo / train_grpo_sd.py, the construction is different:
samples = {
    "prompt_embeds": all_prompts_embed,
    "timesteps": timesteps[:, :-1],
    "latents": all_latents[:, :-1][:, :-1],      # each entry is the latent before timestep t
    "next_latents": all_latents[:, 1:][:, :-1],  # each entry is the latent after timestep t
    "log_probs": all_log_probs[:, :-1],
    "rewards": all_rewards,
}
Could you please explain the reasoning behind this additional slicing ([:, :-1]) and the use of both [:, :-1][:, :-1] for latents and [:, 1:][:, :-1] for next_latents?
Was this change made for stability, alignment of time-step indexing, or to handle an off-by-one issue in the diffusion step alignment?
Understanding the rationale would help clarify the correspondence between latent trajectories and timesteps in GRPO-based diffusion training.

**Comment by XueZeyue** [comment]
Because I want to align the settings with flow-based GRPO training.
I ignore the last timestep in flow-based GRPO to prevent the noisy images, a recent paper also discuss this problem. (https://arxiv.org/pdf/2509.05952)

**Comment by ZongxinZhu1** [comment]
> Because I want to align the settings with flow-based GRPO training. I ignore the last timestep in flow-based GRPO to prevent the noisy images, a recent paper also discuss this problem. (https://arxiv.org/pdf/2509.05952)

Okay, thanks a lot for your reply! 😊!

 I have another question I hope you could clarify: I tried not ignoring the last latent, but after training for one epoch, the generated images turned out completely black. 

Could you help me understand why this happens?

**Comment by XueZeyue** [comment]
I suspect there might be an issue in the DDIM solver where a division by zero occurs in the final training step.

**Comment by ZongxinZhu1** [comment]
Hi XueZeyue,

Thanks for your suggestion! I noticed something interesting on my side when working on **medical image generation:**

When I **train without LoRA**, in the first epoch, the ratio tensor becomes [None, None, None, None] and grad_fn=<ExpBackward0>.

However, when I **switch to LoRA training**, the ratio values are normal, and training proceeds without issues.

This makes me suspect that the problem might be related to NaNs or very small values in the predicted noise (noise_pred) during the early steps of full fine-tuning. Specifically, the ratio = exp(log_prob - old_log_prob) could become undefined if log_prob contains NaN or None, which might happen if:

Numerical instability in the DDIM solver: a division by zero or extremely small variance in the final step could make log_prob undefined.

Unstable UNet outputs in full fine-tuning: during medical image generation, full fine-tuning may produce very large or unstable noise predictions in the first epoch, while LoRA keeps the pretrained weights mostly frozen, stabilizing the outputs.

Gradient explosion: without LoRA, all UNet parameters are updated from scratch, which can trigger very large log-prob differences and cause exp() to overflow to NaN.

**Comment by XueZeyue** [comment]
OK, many thanks for your thoughts and exploration!
